proptest-macro 0.3 / proptest 1.8 breaks mutable property arguments

```rust
#[property_test]
fn extend(mut a: Vec<f64>) {
    a.push(1.0)
}
```
used to work fine, but with this latest pair of updates, it gives the following error
```
error[E0596]: cannot borrow `a` as mutable, as it is not declared as mutable
  --> bin/tests/stats_test.rs:14:5
   |
14 |     a.push(1.0)
   |     ^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
13 | fn extend(mut mut a: Vec<f64>) {
   |               +++
```

<details>
<summary>The macro expands to the following</summary>

```rust
#[test]
fn extend() {
    #[derive(Debug)]
    struct ExtendArgs {
        a: Vec<f64>,
    }
    impl ::proptest::prelude::Arbitrary for ExtendArgs {
        type Parameters = ();
        type Strategy = ::proptest::strategy::Map<
            ::proptest::arbitrary::StrategyFor<(Vec<f64>,)>,
            fn((Vec<f64>,)) -> Self,
        >;
        fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
            use ::proptest::strategy::Strategy;
            ::proptest::prelude::any::<(Vec<f64>,)>().prop_map(|(a,)| Self { a })
        }
    }
    let config = ::proptest::test_runner::Config {
        test_name: Some("module::path::$ test_name"),
        source_file: Some(file!()),
        ..::proptest::test_runner::Config::default()
    };
    let mut runner = ::proptest::test_runner::TestRunner::new(config);
    let result = runner.run(
        &::proptest::strategy::Strategy::prop_map(
            ::proptest::prelude::any::<ExtendArgs>(),
            |values| ::proptest::sugar::NamedArguments("ExtendArgs", values),
        ),
        |::proptest::sugar::NamedArguments(_, ExtendArgs { a })| {
            let result = { a.push(1.0) };
            let _ = result;
            Ok(())
        },
    );
    match result {
        Ok(()) => {}

        Err(e) => {
            #[cold]
            #[track_caller]
            #[inline(never)]
            #[rustc_const_panic_str]
            #[rustc_do_not_const_check]
            const fn panic_cold_display<T: core::fmt::Display>(arg: &T) -> ! {
                core::panicking::panic_display(arg)
            }
            panic_cold_display(&e);
        }
    }
}
```
</details>
<details>
<summary>The previous version was</summary>

```rust
#[test]
fn extend() {
    #[derive(Debug)]
    struct ExtendArgs {
        field0: Vec<f64>,
    }
    impl ::proptest::prelude::Arbitrary for ExtendArgs {
        type Parameters = ();
        type Strategy = ::proptest::strategy::Map<
            ::proptest::arbitrary::StrategyFor<(Vec<f64>,)>,
            fn((Vec<f64>,)) -> Self,
        >;
        fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
            use ::proptest::strategy::Strategy;
            ::proptest::prelude::any::<(Vec<f64>,)>().prop_map(|(field0,)| Self { field0 })
        }
    }
    let config = ::proptest::test_runner::Config {
        test_name: Some("module::path::$ test_name"),
        source_file: Some(file!()),
        ..::proptest::test_runner::Config::default()
    };
    let mut runner = ::proptest::test_runner::TestRunner::new(config);
    let result = runner.run(
        &::proptest::strategy::Strategy::prop_map(
            ::proptest::prelude::any::<ExtendArgs>(),
            |values| ::proptest::sugar::NamedArguments("ExtendArgs", values),
        ),
        |::proptest::sugar::NamedArguments(_, ExtendArgs { field0: mut a })| {
            let result = { a.push(1.0) };
            let _ = result;
            Ok(())
        },
    );
    match result {
        Ok(()) => {}

        Err(e) => {
            #[cold]
            #[track_caller]
            #[inline(never)]
            #[rustc_const_panic_str]
            #[rustc_do_not_const_check]
            const fn panic_cold_display<T: core::fmt::Display>(arg: &T) -> ! {
                core::panicking::panic_display(arg)
            }
            panic_cold_display(&e);
        }
    }
}
```
</details>

Diff:
```diff
--- #<buffer *old*>
+++ #<buffer *new*>
@@ -2,7 +2,7 @@
 fn extend() {
     #[derive(Debug)]
     struct ExtendArgs {
-        field0: Vec<f64>,
+        a: Vec<f64>,
     }
     impl ::proptest::prelude::Arbitrary for ExtendArgs {
         type Parameters = ();
@@ -12,7 +12,7 @@
         >;
         fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
             use ::proptest::strategy::Strategy;
-            ::proptest::prelude::any::<(Vec<f64>,)>().prop_map(|(field0,)| Self { field0 })
+            ::proptest::prelude::any::<(Vec<f64>,)>().prop_map(|(a,)| Self { a })
         }
     }
     let config = ::proptest::test_runner::Config {
@@ -26,7 +26,7 @@
             ::proptest::prelude::any::<ExtendArgs>(),
             |values| ::proptest::sugar::NamedArguments("ExtendArgs", values),
         ),
-        |::proptest::sugar::NamedArguments(_, ExtendArgs { field0: mut a })| {
+        |::proptest::sugar::NamedArguments(_, ExtendArgs { a })| {
             let result = { a.push(1.0) };
             let _ = result;
             Ok(())
```

The `mut` has gone.

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -48,6 +48,8 @@ jobs:
         run: cd proptest && cargo test --verbose
       - name: Run macro tests
         run: cd proptest-macro && cargo test --verbose
+      - name: Run macro integration tests
+        run: cargo test --verbose --test attr_macro --features attr-macro
       - name: Build coverage no-default-features
         if: ${{ matrix.build == 'stable' }}
         env:

--- a/proptest-macro/src/property_test/codegen/test_body.rs
+++ b/proptest-macro/src/property_test/codegen/test_body.rs
@@ -30,7 +30,14 @@ pub(super) fn body(
         // shorthand struct initialization.
 
         match pat {
-            Pat::Ident(_) => quote!(#field_name,),
+            // We need to make sure to handle any mutability modifiers here, i.e. if the user wrote
+            // `mut x: i32`, we have to generate `mut x`, not `x: mut x`
+            //
+            // See https://github.com/proptest-rs/proptest/issues/601
+            Pat::Ident(i) => match i.mutability {
+                Some(mutability) => quote!(#mutability #field_name,),
+                None => quote!(#field_name,),
+            },
             _ => quote!(#field_name: #pat,),
         }
     });

--- a/proptest-macro/src/property_test/utils.rs
+++ b/proptest-macro/src/property_test/utils.rs
@@ -1,9 +1,4 @@
-use core::mem::replace;
-
-use syn::{
-    punctuated::Punctuated, AttrStyle, Attribute, Expr, FnArg, ItemFn, Meta,
-    PatType,
-};
+use syn::{AttrStyle, Attribute, Expr, FnArg, ItemFn, Meta, PatType};
 
 /// A parsed argument, with an optional custom strategy
 pub struct Argument {
@@ -15,7 +10,7 @@ pub struct Argument {
 ///
 /// Panics on any invalid function
 pub fn strip_args(mut f: ItemFn) -> (ItemFn, Vec<Argument>) {
-    let args = replace(&mut f.sig.inputs, Punctuated::new());
+    let args = std::mem::take(&mut f.sig.inputs);
     let args = args
         .into_iter()
         .map(|arg| match arg {

--- a/proptest/tests/attr_macro.rs
+++ b/proptest/tests/attr_macro.rs
@@ -1,0 +1,11 @@
+/// Regression for https://github.com/proptest-rs/proptest/issues/601
+#[cfg(feature = "attr-macro")]
+#[proptest::property_test]
+fn attr_macro_does_not_clobber_mutability(mut x: i32, (mut y, _z): (i32, i32)) {
+    let _suppress_unused_warning = x == y;
+
+    x = 0;
+    y = 0;
+    assert_eq!(x, y);
+}
+

